homeassistant/components/hyperion/switch.py

```diff
@@ -0,0 +1,206 @@
+    return True
+
+
+class HyperionComponentSwitch(SwitchEntity):
+    """ComponentBinarySwitch switch class."""
+
+    def __init__(
+        self,
+        unique_id: str,
+        name: str,
+        component_name: str,
+        hyperion_client: client.HyperionClient,
+    ) -> None:
+        """Initialize the switch."""
+        self._unique_id = unique_id
+        self._name = name
+        self._component_name = component_name
+        self._client = hyperion_client
+        self._client_callbacks = {
+            f"{KEY_COMPONENTS}-{KEY_UPDATE}": self._update_components
+        }
+
+    @property
+    def should_poll(self) -> bool:
+        """Return whether or not this entity should be polled."""
+        return False
+
+    @property
+    def entity_registry_enabled_default(self) -> bool:
+        """Whether or not the entity is enabled by default."""
+        # These component controls are for advanced users and are disabled by default.
+        return False
+
+    @property
+    def unique_id(self) -> str:
+        """Return a unique id for this instance."""
+        return self._unique_id
+
+    @property
+    def name(self) -> str:
+        """Return the name of the switch."""
+        return self._name
+
+    @property
+    def is_on(self) -> bool:
+        """Return true if the switch is on."""
+        for component in self._client.components:
+            if component[KEY_NAME] == self._component_name:
+                return bool(component.setdefault(KEY_ENABLED, False))
+        return False
+
+    @property
+    def available(self) -> bool:
+        """Return server availability."""
+        return bool(self._client.has_loaded_state)
+
+    async def _async_send_set_component(self, value: bool) -> None:
+        """Send a component control request."""
+        await self._client.async_send_set_component(
+            **{
+                KEY_COMPONENTSTATE: {
+                    KEY_COMPONENT: self._component_name,
+                    KEY_STATE: value,
+                }
+            }
+        )
+
+    # pylint: disable=unused-argument
+    async def async_turn_on(self, **kwargs: Any) -> None:
+        """Turn on the switch."""
+        await self._async_send_set_component(True)
+
+    # pylint: disable=unused-argument
+    async def async_turn_off(self, **kwargs: Any) -> None:
+        """Turn off the switch."""
+        await self._async_send_set_component(False)
+
+    def _update_components(self, _: Optional[Dict[str, Any]] = None) -> None:
+        """Update Hyperion components."""
+        self.async_write_ha_state()
+
+    async def async_added_to_hass(self) -> None:
+        """Register callbacks when entity added to hass."""
+        assert self.hass
+        self.async_on_remove(
+            async_dispatcher_connect(
+                self.hass,
+                SIGNAL_ENTITY_REMOVE.format(self._unique_id),
+                self.async_remove,
```

Comment:
This removes the entity. Is that necessary?

Home Assistant already removes entities if the corresponding entity registry entry is removed.

Reply:
Thanks. I think it's necessary. When an 'instance' is stopped on the Hyperion server, the entity is removed from HA but intentionally maintained in the registry (e.g. to preserve the entity_id). When it is *deleted* from the Hyperion server entirely, we then go one step further and remove  it from the registry also. [Relevant Code](https://github.com/home-assistant/core/pull/45410/files#diff-55bc4ad3c32b682a8fcb0e5f1f338118e1c0b94ceb6469aa95e5cbd7a5acbdc0R264). 

Let me know if I'm misunderstanding something.

Reply:
Ah ok. Yeah that makes sense then. I guess I was confused because the signal is `SIGNAL_ENTITY_REMOVE` which is the action is meant to be taken, instead of describing the thing that just happened. 

Right now removing entities will result in the entity being removed. Once we merge #45528, they will written as `unavailable` states in the state machine, just like they are when Home Assistant is restarted. 

For now it's fine 👍